### ISSUE TITLE
CONSTRUCTION [RESEARCH]: StructuralPRM — process reward model from call graph requiring zero training data

### ISSUE THREAD

**Issue by nateschmiedehaus** [nateschmiedehaus/LiBrainian#331]
# CONSTRUCTION [RESEARCH]: StructuralPRM — process reward model from call graph (zero training data)

## Problem Statement

Process Reward Models (PRMs) for code dramatically outperform outcome reward models at large sampling budgets (ICLR 2024). But training a PRM requires expensive step-level annotations: either human annotators label thousands of intermediate reasoning steps as correct/incorrect, or execution-based oracles verify each step by running code. Both approaches are costly.

This issue proposes a novel PRM that requires **zero training data** by using librainian's call graph, type index, and semantic function descriptions as the step verifier. Every agent claim about the codebase ("function X has callers Y and Z") is checkable in O(1) against the index. This is the first code PRM that generates labeled training examples as a side effect of normal agent operation.

Note: #291 covers the engineering of StructuralPRM as an inference harness. This issue covers the **research contribution**: the formal framing, the zero-training-data theorem, the `StructuralPRM.verify()` Construction interface, and the venue/publication path.

## Formal Background

**Process Reward Models** (Lightman et al., ICLR 2024, "Let's Verify Step by Step"): a PRM scores each intermediate reasoning step in a chain-of-thought. At large N (many candidate solutions), PRM-guided best-of-N dramatically outperforms outcome reward models. Training requires step-level annotations.

**Math-Shepherd** (Wang et al., arXiv 2312.08935): reduces PRM annotation cost by auto-generating step labels via *execution-based verification* — run the code at each step, check if tests pass. Still requires execution infrastructure and does not generalize to steps that cannot be run.

**The structural verification insight**: for codebase reasoning, librainian's call graph and type index can verify structural claims *without execution*. Claims of the form:
- "Function F exists at path P" → verify against file index, O(1)
- "F's callers are X and Y" → verify against call graph, O(degree)
- "F's signature is `(a: string) => Promise<void>`" → verify against type index, O(1)
- "F imports module M" → verify against import graph, O(1)

are all checkable in constant or near-constant time against the existing index. No code execution needed.

**The zero-training-data theorem**: if an agent reasons about 5,295 indexed functions and generates N reasoning steps per function, each step is labeled automatically (grounded vs. contradicted) as a side effect of normal `withProofTrace` operation. At N=3 steps/function, this generates ~15,000 labeled examples from index data alone — before any agent has ever been deployed.

Sources:
- Lightman, H., et al. (2024). *Let's Verify Step by Step*. ICLR 2024.
- Wang, P., et al. (2023). *Math-Shepherd: Verify and Reinforce LLMs Step-by-step without Human Annotations*. arXiv:2312.08935.
- See also: #291 (WAVE6 [HARNESS]: StructuralPRM) for the engineering implementation.

## Proposed Interface (TypeScript)

```typescript
// The StructuralPRM — verifies a single reasoning step against the index
interface StructuralPRM {
  /**
   * Verify a natural-language reasoning step about the codebase.
   *
   * @param reasoningStep - A string from the agent's CoT trace
   * @param sessionCtx    - The librarian context for ground-truth lookup
   * @returns ClaimVerdict: the step's structural correctness status
   */
  verify(
    reasoningStep: string,
    sessionCtx: LibrarianContext
  ): Promise<ClaimVerdict>;

  /**
   * Score an entire reasoning trace.
   * Returns per-step verdicts and an aggregate process reward score.
   */
  scoreTrace(
    trace: string[],
    sessionCtx: LibrarianContext
  ): Promise<TraceScore>;
}

// The verdict for a single step
type ClaimVerdict =
  | { status: 'grounded';       evidenceRef: EvidenceId;  confidence: number }
  | { status: 'contradicted';   defeater: string;          confidence: number }
  | { status: 'uncheckable';    reason: string }  // claim makes no structural assertions

// Aggregate score for a full trace
interface TraceScore {
  perStep: ClaimVerdict[];
  processRewardScore: number;    // aggregate score in [0, 1]
  groundedFraction: number;      // fraction of checkable steps that were grounded
  labeledExamples: LabeledStep[]; // side-effect: training data for future PRM training
}

// Labeled training example generated as a side effect
interface LabeledStep {
  step: string;
  label: 'correct' | 'incorrect';
  groundTruth: string;   // what the index actually says
  entityRefs: string[];  // which entities the step referenced
}

// Constructor: factory function
declare function createStructuralPRM(options?: {
  /** Minimum confidence to mark a step as 'grounded'. Default: 0.8 */
  groundingThreshold?: number;
  /** Whether to emit labeled examples to a training corpus. Default: false */
  emitTrainingData?: boolean;
  /** Path to append labeled examples to. Required if emitTrainingData: true */
  trainingCorpusPath?: string;
}): StructuralPRM;
```

## Research Contribution Summary

**Title**: StructuralPRM: A Zero-Training-Data Process Reward Model for Code Agents Using Codebase Knowledge Graphs

**The novel claim**: A codebase knowledge graph (call graph + type index + semantic function index) serves as a structural oracle for PRM training data generation, requiring zero human annotations and zero code execution at labeling time. This is categorically different from execution-based methods (Math-Shepherd) because:
1. Labeling is O(1) per claim, not O(test suite runtime)
2. Labels are available before any agent run — the training corpus pre-exists
3. Incorrect structural claims are labeled with ground truth (what the index actually says), providing high-quality negative examples

**Immediate value**: The 5,295 functions in librainian's current index generate a labeled training corpus as a side effect of `withProofTrace` operation. At 3 reasoning steps per function: ~15,000 labeled examples with zero annotation cost.

**Research venue**: PLDI or ICML (ML Systems track)

## Acceptance Criteria

- [ ] `StructuralPRM.verify(step, ctx)` correctly identifies grounded steps (claims verifiable against the call graph) and returns `{ status: 'grounded' }` for a step containing a correct claim about a function's callers.
- [ ] `verify(step, ctx)` returns `{ status: 'contradicted' }` for a step stating a caller relationship that does not exist in the call graph, with the `defeater` field populated with the correct caller count.
- [ ] `verify(step, ctx)` returns `{ status: 'uncheckable' }` for a step making no structural assertions ("I should refactor this function").
- [ ] `scoreTrace(trace, ctx)` returns a `processRewardScore` that correlates positively with the fraction of steps that are grounded — verified on a held-out set of 50 manually-labeled reasoning traces.
- [ ] When `emitTrainingData: true`, `scoreTrace()` appends `LabeledStep` objects to the specified corpus path in JSONL format.
- [ ] Integration with `withProofTrace` (issue #[5-ops]): a Construction wrapped with `withProofTrace` uses `StructuralPRM.verify` as the step verifier implementation.
- [ ] Cross-reference #291 in implementation: `StructuralPRM` is the shared backing implementation for both the HARNESS (inference-time scoring) and the OPERATOR (training data generation).
